### PR TITLE
fix(app): prevent OverflowError from addresses above 2^63 in memory map

### LINKED ISSUE #73: Memory map dialog Qt overflow error
**Describe the bug**
The memory map widget causes an error to get reported in the console.

**To Reproduce**
1. Open the App
2. Attach to any process
3. Open "Memory Map"

**Expected behavior**
No error is reported in the console.

**System (please complete the following information):**
 - OS: NixOS (Linux)
 - Python Version 3.14

**Additional context**
```
OverflowError
OverflowError: int too big to convert
OverflowError
Traceback (most recent call last):
  File "/nix/store/i6ldivv97db8sxzx9dy93s9v98fw9q4h-python3.14-PyMemoryEditor-2.0.1/lib/python3.14/site-packages/PyMemoryEditor/app/_auto_refresh_dialog.py", line 116, in _handle_ready
    self._on_data_ready(data)
    ~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/nix/store/i6ldivv97db8sxzx9dy93s9v98fw9q4h-python3.14-PyMemoryEditor-2.0.1/lib/python3.14/site-packages/PyMemoryEditor/app/memory_map_dialog.py", line 355, in _on_data_ready
    self._populate()
    ~~~~~~~~~~~~~~^^
  File "/nix/store/i6ldivv97db8sxzx9dy93s9v98fw9q4h-python3.14-PyMemoryEditor-2.0.1/lib/python3.14/site-packages/PyMemoryEditor/app/memory_map_dialog.py", line 394, in _populate
    addr_item.setData(addr, Qt.UserRole)
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
OverflowError
```

### PR DESCRIPTION
**Pull request by cromachina**
**Why is this PR necessary, what does it do?**

Fixes an error that appears when viewing the memory map (and also potentially the module viewer). The error prevents the lists from being fully populated. The error can also potentially segfault the program, at least when shiboken stuff runs.

Also includes a minor cosmetic change to make comparing the addresses in a list easier (monospace font for the addresses).

**Checklist (complete all items)**:

- [✅] Added tests as necessary.
- [✅] There is no breaking change for existing features.

**References:**

Fixes #73

**Notes:**

No notes to be shared.

### CODE REVIEW

**Comment by cromachina** [comment]
The "Lint PR title" check doesn't check case insensitive, even though github generates titles from branches automatically with casing.

**Comment by JeanExtreme002** [comment]
Hi, @cromachina! Thanks for contributing. 

There are some issues that must be fixed before merging this PR. I'll post the reviews here.

File: PyMemoryEditor/app/_widgets.py [hunk 1]
```diff
@@ -66,6 +66,22 @@ class NumericItem(QStandardItem):
...
+        else:
+            super().setData(value, role)
+
+    def data(self, role: int = Qt.UserRole + 1) -> int:
+        if role >= Qt.UserRole:
+            return self._data
+        else:
+            return super().data(role)
 
     def __lt__(self, other):
```
Comment (JeanExtreme002) [inline PyMemoryEditor/app/_widgets.py]: **This fallback segfaults the app.** `QStandardItem::operator<` is virtual, so `super().__lt__(other)` on line 90 dispatches straight back into this override and recurses until the stack blows up — a crash, not an exception, once the C++ sort is driving the comparison.

It's reachable today: `open_process_dialog.py:309` calls `setData(mem, Qt.UserRole)` unconditionally, and `mem` is `None` for processes whose memory couldn't be read. Sorting the Memory column crashes (reproduced: exit 139). The new `_data = 0` default happens to hide it in the Threads/Priority column, where `setData` is conditional — but not here.

Fix is to compare labels instead of delegating up:

```python
        except (TypeError, ValueError):
            return self.text() < other.text()
```

File: PyMemoryEditor/app/_widgets.py [hunk 2]
```diff
@@ -66,6 +66,22 @@ class NumericItem(QStandardItem):
     Used by columns showing formatted numbers (sizes, addresses, PIDs) so the
     table sorts by the underlying value rather than the lexical label.
     """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._data = 0
```
Comment (JeanExtreme002) [inline PyMemoryEditor/app/_widgets.py]: A single slot makes every role `>= UserRole` alias each other: `setData(111, Qt.UserRole)` then `setData(222, Qt.UserRole + 1)` leaves both roles reading `222`. Plain `QStandardItem` stores one value per role. Nothing uses `UserRole + 1` today, so it's latent — but a dict keeps the native semantics for free:

```python
self._user_data: Dict[int, Any] = {}          # __init__
self._user_data[int(role)] = value            # setData
return self._user_data.get(int(role))         # data
```

File: PyMemoryEditor/app/_widgets.py [hunk 3]
```diff
@@ -66,6 +66,22 @@ class NumericItem(QStandardItem):
...
+
+    def setData(self, value: int, role: int = Qt.UserRole + 1):
```
Comment (JeanExtreme002) [inline PyMemoryEditor/app/_widgets.py]: `value: int` here and `-> int` on line 80 aren't accurate — `setFont`, `setToolTip` and `setForeground` all dispatch through this override with `QFont`/`str`/`QColor` before landing in the `super()` branch. `Any` would be truer.

File: PyMemoryEditor/app/_widgets.py [hunk 4]
```diff
@@ -66,6 +66,22 @@ class NumericItem(QStandardItem):
     Used by columns showing formatted numbers (sizes, addresses, PIDs) so the
     table sorts by the underlying value rather than the lexical label.
     """
```
Comment (JeanExtreme002) [inline PyMemoryEditor/app/_widgets.py:85]: Worth a line in the docstring on why the overrides exist (the `QVariant` `qint64` cap vs. the `[vsyscall]` page), matching the "why" comments elsewhere in this module — otherwise this looks removable later.

Worth stating the constraint it introduces too: the payload never reaches the C++ model, so it has to be read as `item.data(role)` and never `model.data(index, role)` — the latter converts back to a `QVariant` and overflows again (confirmed).

File: PyMemoryEditor/app/memory_map_dialog.py [hunk 5]
```diff
@@ -391,6 +391,7 @@ def _populate(self) -> None:
             shown += 1
 
             addr_item = NumericItem(f"0x{addr:016X}")
+            addr_item.setFont(QFontDatabase.systemFont(QFontDatabase.SystemFont.FixedFont))
```
Comment (JeanExtreme002) [inline PyMemoryEditor/app/memory_map_dialog.py]: The app already standardises on `QFont("Menlo, Consolas, Courier New", 10)` — `_MONO` at `pointer_scan_dialog.py:61`, plus `memory_viewer_dialog`, `log_console_dialog` and `pointer_chain_dialog`. `systemFont(FixedFont)` resolves to a different family at a platform-dependent size (9pt here), so this column ends up next to the 10pt Menlo of `_size_edit` (line 268 of this same file) and out of step with the Pointer Scan table.

Suggest promoting the family to `_widgets.py` (next to `NumericItem`, which these columns already import) and building the `QFont` once above the loop rather than per row.

File: PyMemoryEditor/app/modules_dialog.py [hunk 6]
```diff
@@ -167,6 +167,7 @@ def _apply_filter(self) -> None:
 
             base = int(module.base_address)
             base_item = NumericItem(f"0x{base:016X}")
+            base_item.setFont(QFontDatabase.systemFont(QFontDatabase.SystemFont.FixedFont))
```
Comment (JeanExtreme002) [inline PyMemoryEditor/app/modules_dialog.py]: Same as the memory map: prefer the app's existing monospace stack, built once above the loop.

**Review by JeanExtreme002 — COMMENTED** [review]
Good catch on the root cause. Verified locally: regular addresses (`0x7ffd…`) go through `setData` fine — only values above 2**63 overflow, which is exactly `[vsyscall]` at `0xffffffffff600000` on Linux x86-64. The Python-side payload is the right call, and it also covers `pointer_scan_dialog.py:687`, which had the same latent bug.

Three things before merge, inline. The first one is a crash, the other two are smaller.

One more: the checklist says tests were added, but there are none in the diff and `_widgets.py` has no test file yet. `tests/app/` + `QT_QPA_PLATFORM=offscreen` already make this cheap to cover — a round trip with `0xffffffffff600000` and a sort over rows with a missing payload would pin both regressions.

File: PyMemoryEditor/app/memory_map_dialog.py (hunk 5 above)
Comment (cromachina) [reply PyMemoryEditor/app/memory_map_dialog.py]: Maybe all of those other files should also reference `_MONO` either from where it's defined or some `common.py`/`shared.py` file?

Comment (cromachina) [reply PyMemoryEditor/app/memory_map_dialog.py]: Ah, in `_widgets.py`, right.

File: PyMemoryEditor/app/_widgets.py [hunk 7]
```diff
@@ -65,13 +65,41 @@ class NumericItem(QStandardItem):
...
+        else:
+            super().setData(value, role)
+
+    def data(self, role: int = Qt.UserRole) -> Any:
+        if role >= Qt.UserRole:
+            return self._user_data.get(int(role))
+        else:
+            return super().data(role)
+
+    def text(self) -> str:
```
Comment (JeanExtreme002) [inline PyMemoryEditor/app/_widgets.py]: **This override should come out.** `dict.get()` never raises `KeyError`, it returns `None` — so the `except` branch is dead and `super().text()` is unreachable. What actually happens:

```
label shown in the table (DisplayRole) : 0x00007FFDABCD1234
item.text()                            : 140727485796916
no payload -> text()                   : 'None'
```

Two consequences:

1. `text()` stops matching what the table displays — it returns the raw payload instead of the formatted label.
2. It cancels out the `__lt__` fix below: the fallback calls `self.text()`, which now yields `'None'` on both sides, so `NumericItem("aaa") < NumericItem("bbb")` returns `False`. No more segfault (the important part), but the lexical comparison never happens.

Nothing in the UI breaks today — `QStandardItem::text()` isn't virtual, so C++ never calls this, and the `.text()` calls in the dialogs are all on plain `QStandardItem` columns. Still a trap worth removing: `__lt__` wants the label, and `super().text()` already gives it.

File: PyMemoryEditor/app/memory_map_dialog.py [hunk 8]
```diff
@@ -38,6 +38,7 @@
 
 from ._auto_refresh_dialog import AutoRefreshTableDialog
 from ._widgets import NumericItem
+from .pointer_scan_dialog import _MONO
```
Comment (JeanExtreme002) [inline PyMemoryEditor/app/memory_map_dialog.py]: Importing a private name from another dialog module points the dependency the wrong way — the memory map now pulls in the whole pointer scan module for one string, and `modules_dialog.py:43` does the same.

`_widgets.py` is where `NumericItem` already lives and exists precisely for helpers "that previously appeared duplicated across several dialog modules". Moving the constant there (and having `pointer_scan_dialog` import it instead of redeclaring it) keeps all three dialogs pointing at one place.

File: tests/app/test_app_smoke.py [hunk 9]
```diff
@@ -162,3 +162,34 @@ def test_pointer_scan_dialog_constructs_and_prefills(qtbot):
...
+    item = _widgets.NumericItem()
+    item.setData(unsigned_64bit_max)
+    assert item.data() == unsigned_64bit_max
+
+    item2 = _widgets.NumericItem()
+    item2.setData(big_number)
+    assert item2.data() == big_number
+
+    # Segmentation fault regression (recursive stack overflow).
+    item < item2
```
Comment (JeanExtreme002) [inline tests/app/test_app_smoke.py]: A few gaps here:

- `item < item2` and the two `item < item3` lines have no assertion — they only prove nothing crashed. Worth asserting the expected order so a future regression fails loudly instead of silently reordering rows.
- No coverage for the role collision that was just fixed (`setData(111, Qt.UserRole)` / `setData(222, Qt.UserRole + 1)` must stay distinct).
- No coverage for `model.sort()` over rows with a missing payload — that's the case that actually segfaulted, and comparing two items directly doesn't exercise the C++ sort path.
- These are `_widgets` tests living in the smoke-test file, gated behind `pytest-qt` they don't need (a `QApplication` is enough). A `tests/app/test_app_widgets.py` would keep them running even without `pytest-qt` installed.

**Review by JeanExtreme002 — COMMENTED** [review]
Thanks for the quick turnaround — roles, type hints and the sort recursion are all sorted (verified: sorting a column with `None` payloads no longer crashes, and `UserRole` / `UserRole + 1` stay distinct). Resolved those threads.

One new issue from the `text()` override, plus two smaller things, inline.

For the record on the docstring: the cap isn't really about unsigned support — a plain `QStandardItem` takes `0x7FFDABCD1234` fine and only breaks past 2**63, i.e. `QVariant`'s `qint64`. Your `2**128` case shows the same edge. Also worth stating the constraint the workaround introduces: the payload never reaches the C++ model, so it must be read as `item.data(role)` and never `model.data(index, role)` — that path converts back to a `QVariant` and overflows again (confirmed locally).

File: PyMemoryEditor/app/_widgets.py (hunk 1 above)
Comment (JeanExtreme002) [reply PyMemoryEditor/app/_widgets.py]: The recursion itself is gone — confirmed, sorting a column with `None` payloads no longer crashes. Leaving this open because the new `text()` override sends the fallback to `'None'` on both sides, so the lexical comparison it's supposed to do never runs. Removing that override closes this one out.

File: PyMemoryEditor/app/_widgets.py [hunk 10]
```diff
@@ -65,13 +65,41 @@ class NumericItem(QStandardItem):
...
+        try:
+            return str(self.data())
+        except (KeyError):
+            return super().text()
```
Comment (JeanExtreme002) [inline PyMemoryEditor/app/_widgets.py]: Suggestion for the point above — dropping the override restores `QStandardItem.text()`, which is what `__lt__` needs for its fallback:

```suggestion
```

File: tests/app/test_app_smoke.py [hunk 11]
```diff
@@ -162,3 +162,34 @@ def test_pointer_scan_dialog_constructs_and_prefills(qtbot):
...
+
+    item3 = _widgets.NumericItem()
+    item3.setData('123456')
+    item < item3
+
+    item3.setData('hello world')
+    item < item3
```
Comment (JeanExtreme002) [inline tests/app/test_app_smoke.py]: Suggestion for the gaps above — assertions on the comparisons, plus the role collision and the `model.sort()` path:

```suggestion
    from PySide6.QtCore import Qt
    from PySide6.QtGui import QStandardItemModel

    # Non-numeric payloads must fall back to the labels instead of recursing
    # into QStandardItem::operator< (that recursion segfaulted mid-sort).
    assert item < item2

    item3 = _widgets.NumericItem('aaa')
    item3.setData('hello world')
    item4 = _widgets.NumericItem('bbb')
    item4.setData('hello world')
    assert item3 < item4
    assert not (item4 < item3)

    # Distinct user roles must not share a slot.
    item5 = _widgets.NumericItem()
    item5.setData(111, Qt.UserRole)
    item5.setData(222, Qt.UserRole + 1)
    assert item5.data(Qt.UserRole) == 111
    assert item5.data(Qt.UserRole + 1) == 222

    # The path that actually crashed: the C++ sort driving the comparisons over
    # a column mixing payloads and None (the process picker's memory column).
    model = QStandardItemModel()
    for label, payload in (('120 MB', 120), ('-', None), ('8 MB', 8), ('-', None)):
        row_item = _widgets.NumericItem(label)
        row_item.setData(payload, Qt.UserRole)
        model.appendRow([row_item])
    model.sort(0, Qt.AscendingOrder)
    order = [model.item(row, 0).data(Qt.DisplayRole) for row in range(model.rowCount())]
    assert order.index('8 MB') < order.index('120 MB')
```

The `item3 < item4` pair is what pins the `text()` fix: with the override in place both sides read `'hello world'` and the assertion fails.

One thing a suggestion can't do — moving this to `tests/app/test_app_widgets.py`. Worth doing so it isn't gated behind the `pytest-qt` skip it doesn't need (none of it uses `qtbot`, just a `QApplication`).

**Review by JeanExtreme002 — COMMENTED** [review]
Adding applicable suggestions for the two open points above. Both verified locally: with them the app suite is 83 passed and flake8 is clean; putting the `text()` override back makes the new assertions fail.

**Comment by JeanExtreme002** [comment]
I went ahead and pushed the two suggestions to this branch (`a82087b`) since the branch allows maintainer edits — feel free to amend or revert if you'd rather do it yourself.

- Dropped the `text()` override. Its `except KeyError` was unreachable (`dict.get()` returns `None`), so `text()` was returning the raw payload instead of the label, and the `__lt__` fallback was comparing `'None'` to `'None'` instead of the labels.
- Added assertions to the NumericItem test, plus coverage for the role collision and for `model.sort()` over a column mixing payloads with `None` — the case that actually segfaulted, which comparing two items directly doesn't reach.

Verified locally: 406 passed, 78 skipped, flake8 clean.

Still open: the `model.data()` note in the docstring, and `_MONO` living in `pointer_scan_dialog` rather than `_widgets`. Both are minor.

**Comment by JeanExtreme002** [comment]
Pushed the last two as well (`57bb6ad`):

- `MONOSPACE_FAMILY` now lives in `_widgets.py` next to `NumericItem`, and all three dialogs read it from there instead of importing `_MONO` across dialog modules. The memory map's size field was hardcoding the same string, so it uses the constant too.
- The `NumericItem` tests moved to `tests/app/test_app_widgets.py`. They were behind a `pytest-qt` skip they never needed (no `qtbot`, just a `QApplication`), so they run wherever PySide6 is installed.

406 passed, 78 skipped, flake8 clean. All review threads are resolved now — the only thing I left alone is the docstring wording, which is fine as is.

**Comment by JeanExtreme002** [comment]
Last one from me (`9fcfc0a`) — reworded the docstring rationale. A plain `QStandardItem` takes `0x7FFDABCD1234` fine and only breaks past 2**63, so the cap is `QVariant`'s `qint64` rather than unsigned support (your `2**128` case is the same edge). Also noted the constraint the workaround creates: the payload never reaches the C++ model, so reading it back via `model.data(index, role)` hits the same `OverflowError` — worth having written down before someone adds a delegate or a sort role over these columns.

From my side this is ready to merge.

File: PyMemoryEditor/app/_widgets.py (hunk 7 above)
Comment (cromachina) [reply PyMemoryEditor/app/_widgets.py]: Ah, I had the dict using the accessor which throws KeyError before but changed it and forgot.

**Review by JeanExtreme002 — APPROVED** [review]
LGTM 🚀
